File: devicesetup/src/main/java/io/particle/android/sdk/devicesetup/model/ScanResultNetwork.java
package io.particle.android.sdk.devicesetup.model;

import android.net.wifi.ScanResult;

import java.util.Set;

import static io.particle.android.sdk.utils.Py.set;


// FIXME: this naming... is not ideal.
public class ScanResultNetwork implements WifiNetwork {

    private static final Set<String> wifiSecurityTypes = set("WEP", "PSK", "EAP");


    public final ScanResult scanResult;

    public ScanResultNetwork(ScanResult scanResult) {
        this.scanResult = scanResult;
    }

    @Override
    public String getSsid() {
        return scanResult.SSID;
    }

    @Override
    public boolean isSecured() {
        // <sad trombone>
        // this seems like a bad joke of an "API", but this is basically what
        // Android does internally (see: http://goo.gl/GCRIKi)
        for (String securityType : wifiSecurityTypes) {
            if (scanResult.capabilities.contains(securityType)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ScanResultNetwork that = (ScanResultNetwork) o;

        if (getSsid() != null ? !getSsid().equals(that.getSsid()) : that.getSsid() != null)
            return false;

        return true;
    }

    @Override
    public int hashCode() {
        return getSsid() != null ? getSsid().hashCode() : 0;
    }
}

File: devicesetup/src/main/java/io/particle/android/sdk/devicesetup/model/ScanResultNetwork.kt
package io.particle.android.sdk.devicesetup.model

import android.net.wifi.ScanResult

import io.particle.android.sdk.utils.SSID

import io.particle.android.sdk.utils.Py.set


// FIXME: this naming... is not ideal.
class ScanResultNetwork(private val scanResult: ScanResult) : WifiNetwork {
    override val ssid: SSID? = SSID.from(scanResult.SSID)

    override// <sad trombone>
    // this seems like a bad joke of an "API", but this is basically what
    // Android does internally (see: http://goo.gl/GCRIKi)
    val isSecured: Boolean
        get() {
            for (securityType in wifiSecurityTypes) {
                if (scanResult.capabilities.contains(securityType)) {
                    return true
                }
            }
            return false
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as ScanResultNetwork?

        return if (ssid != null) ssid == that!!.ssid else that!!.ssid == null
    }

    override fun hashCode(): Int {
        return ssid?.hashCode() ?: 0
    }

    companion object {

        private val wifiSecurityTypes = set("WEP", "PSK", "EAP")
    }
}
